In [9]:
import numpy as np
import pandas as pd


In [10]:
df = pd.read_csv('dataset.csv')

In [11]:
df.head()

Release Year                             Title Origin/Ethnicity  \
0          1901            Kansas Saloon Smashers         American   
1          1901     Love by the Light of the Moon         American   
2          1901           The Martyred Presidents         American   
3          1901  Terrible Teddy, the Grizzly King         American   
4          1902            Jack and the Beanstalk         American   

                             Director Cast    Genre  \
0                             Unknown  NaN  unknown   
1                             Unknown  NaN  unknown   
2                             Unknown  NaN  unknown   
3                             Unknown  NaN  unknown   
4  George S. Fleming, Edwin S. Porter  NaN  unknown   

                                           Wiki Page  \
0  https://en.wikipedia.org/wiki/Kansas_Saloon_Sm...   
1  https://en.wikipedia.org/wiki/Love_by_the_Ligh...   
2  https://en.wikipedia.org/wiki/The_Martyred_Pre...   
3  https://en.wikipedia.org/wiki/Terrible_Teddy,_...   
4  https://en.wikipedia.org/wiki/Jack_and_the_Bea...   

                                                Plot  
0  A bartender is working at a saloon, serving dr...  
1  The moon, painted with a smiling face hangs ov...  
2  The film, just over a minute long, is composed...  
3  Lasting just 61 seconds and consisting of two ...  
4  The earliest known adaptation of the classic f...

In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34886 entries, 0 to 34885
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   Release Year      34886 non-null  int64 
 1   Title             34886 non-null  object
 2   Origin/Ethnicity  34886 non-null  object
 3   Director          34886 non-null  object
 4   Cast              33464 non-null  object
 5   Genre             34886 non-null  object
 6   Wiki Page         34886 non-null  object
 7   Plot              34886 non-null  object
dtypes: int64(1), object(7)
memory usage: 2.1+ MB


In [13]:
df['Plot'][0]

"A bartender is working at a saloon, serving drinks to customers. After he fills a stereotypically Irish man's bucket with beer, Carrie Nation and her followers burst inside. They assault the Irish man, pulling his hat over his eyes and then dumping the beer over his head. The group then begin wrecking the bar, smashing the fixtures, mirrors, and breaking the cash register. The bartender then sprays seltzer water in Nation's face before a group of policemen appear and order everybody to leave.[1]"

In [14]:
df['Wiki Page'][0]

'https://en.wikipedia.org/wiki/Kansas_Saloon_Smashers'

In [18]:
df['Genre'].isna()

0        False
1        False
2        False
3        False
4        False
         ...  
34881    False
34882    False
34883    False
34884    False
34885    False
Name: Genre, Length: 34886, dtype: bool

In [19]:
df.isna().sum()

Release Year           0
Title                  0
Origin/Ethnicity       0
Director               0
Cast                1422
Genre                  0
Wiki Page              0
Plot                   0
dtype: int64

In [21]:
df.duplicated().sum()

0

In [24]:
df.describe()

Release Year
count  34886.000000
mean    1981.314252
std       27.815174
min     1901.000000
25%     1957.000000
50%     1988.000000
75%     2007.000000
max     2017.000000

In [ ]:

df['Release Year'].value_counts()


Release Year
2013    1021
2014     929
2012     874
2011     858
2010     825
        ... 
1906       3
1905       2
1903       2
1904       1
1902       1
Name: count, Length: 117, dtype: int64

In [28]:

df['Title'].value_counts()


Title
Cinderella               8
The Three Musketeers     8
Treasure Island          7
Hero                     6
Alice in Wonderland      6
                        ..
Leap of Faith            1
A League of Their Own    1
The Lawnmower Man        1
Ladybugs                 1
İstanbul Kırmızısı       1
Name: count, Length: 32432, dtype: int64

In [31]:

df['Origin/Ethnicity'].value_counts()


Origin/Ethnicity
American        17377
British          3670
Bollywood        2931
Tamil            2599
Telugu           1311
Japanese         1188
Malayalam        1095
Hong Kong         791
Canadian          723
Australian        576
South_Korean      522
Chinese           463
Kannada           444
Bengali           306
Russian           232
Marathi           141
Filipino          128
Bangladeshi        87
Punjabi            84
Malaysian          70
Turkish            70
Egyptian           67
Assamese            9
Maldivian           2
Name: count, dtype: int64

In [33]:

df['Director'].value_counts()


Director
Unknown                1124
Michael Curtiz           79
Hanna-Barbera            77
Lloyd Bacon              66
Jules White              63
                       ... 
Brian Clemens             1
Lester James Peries       1
Maurice Hamblin           1
Trevor Wrenn              1
Ferzan Özpetek            1
Name: count, Length: 12593, dtype: int64

In [34]:

df['Cast'].value_counts()


Cast
Tom and Jerry                                                80
Three Stooges                                                56
The Three Stooges                                            47
Looney Tunes                                                 37
Bugs Bunny                                                   36
                                                             ..
Steve Martin, Goldie Hawn, Dana Delany                        1
Nicolas Cage, James Caan, Sarah Jessica Parker                1
Rick Moranis, Marcia Strassman                                1
Macaulay Culkin, Joe Pesci, Daniel Stern                      1
Halit Ergenç, Tuba Büyüküstün, Mehmet Günsür, Nejat İşler     1
Name: count, Length: 32182, dtype: int64

In [35]:
df['Genre'].value_counts()

Genre
unknown                          6083
drama                            5964
comedy                           4379
horror                           1167
action                           1098
                                 ... 
cbc-tv miniseries                   1
bio-drama                           1
national film board docudrama       1
cult drama                          1
horror romantic comedy              1
Name: count, Length: 2265, dtype: int64

In [36]:
df['Wiki Page'].value_counts()

Wiki Page
https://en.wikipedia.org/wiki/Digimon_Adventure_tri.                         5
https://en.wikipedia.org/wiki/Himitsu_Sentai_Gorenger                        5
https://en.wikipedia.org/wiki/Tamayura                                       4
https://en.wikipedia.org/wiki/Mani_Malai                                     4
https://en.wikipedia.org/wiki/Vettaikaaran_(2009_film)                       3
                                                                            ..
https://en.wikipedia.org/wiki/Hot_Cars,_Cold_Facts                           1
https://en.wikipedia.org/wiki/Home_Alone                                     1
https://en.wikipedia.org/wiki/Heart_Condition_(film)                         1
https://en.wikipedia.org/wiki/Havana_(film)                                  1
https://en.wikipedia.org/wiki/%C4%B0stanbul_K%C4%B1rm%C4%B1z%C4%B1s%C4%B1    1
Name: count, Length: 34070, dtype: int64

In [43]:
(df['Plot'].isna() == True).sum()


0

In [44]:
df1 = df.copy()


In [51]:
import numpy as np
df1.replace(['unknown','Unknown',''], np.nan, inplace=True)


In [52]:
df1.isna().sum()

Release Year           0
Title                  2
Origin/Ethnicity       0
Director            1124
Cast                1423
Genre               6083
Wiki Page              0
Plot                   0
dtype: int64

In [60]:
df1.nunique()


Release Year          117
Title               32431
Origin/Ethnicity       24
Director            12592
Cast                32181
Genre                2264
Wiki Page           34070
Plot                33869
dtype: int64

In [61]:
title_missing = df[df1['Title'].isnull()]
print(title_missing)

       Release Year    Title Origin/Ethnicity                      Director  \
16136          2011  Unknown         American            Jaume Collet-Serra   
21110          2011  Unknown          British  Director: Jaume Collet-Serra   

                                                    Cast            Genre  \
16136  Liam Neeson, Diane Kruger, January Jones, Fran...  action thriller   
21110  Director: Jaume Collet-Serra\r\nCast: Liam Nee...          unknown   

                                               Wiki Page  \
16136  https://en.wikipedia.org/wiki/Unknown_(2011_film)   
21110  https://en.wikipedia.org/wiki/Unknown_(2011_film)   

                                                    Plot  
16136  Dr. Martin Harris and his wife Liz arrive in B...  
21110  Dr. Martin Harris and his wife Liz arrive in B...  


In [62]:

df1.dropna(subset=['Title'], inplace=True)

In [63]:
df1.reset_index(drop=True, inplace=True)

In [64]:
df1.isna().sum()

Release Year           0
Title                  0
Origin/Ethnicity       0
Director            1124
Cast                1423
Genre               6082
Wiki Page              0
Plot                   0
dtype: int64

### found that over 1124 entries in Director is null
### over 1423 entries in the cast is null
### over 6083 entries in the genre is null


In [65]:
import re

def clean_plot(text):
    text = text.lower()
    text = re.sub(r"\[\d+\]", "", text)  # remove [1], [2]
    text = re.sub(r"\{\{.*?\}\}", "", text)  # templates
    text = re.sub(r"==.*?==", "", text)  # headings
    text = re.sub(r"\s+", " ", text)
    return text.strip()


In [66]:
df1["clean_plot"] = df1["Plot"].apply(clean_plot)


In [67]:
df1.head()

Release Year                             Title Origin/Ethnicity  \
0          1901            Kansas Saloon Smashers         American   
1          1901     Love by the Light of the Moon         American   
2          1901           The Martyred Presidents         American   
3          1901  Terrible Teddy, the Grizzly King         American   
4          1902            Jack and the Beanstalk         American   

                             Director Cast Genre  \
0                                 NaN  NaN   NaN   
1                                 NaN  NaN   NaN   
2                                 NaN  NaN   NaN   
3                                 NaN  NaN   NaN   
4  George S. Fleming, Edwin S. Porter  NaN   NaN   

                                           Wiki Page  \
0  https://en.wikipedia.org/wiki/Kansas_Saloon_Sm...   
1  https://en.wikipedia.org/wiki/Love_by_the_Ligh...   
2  https://en.wikipedia.org/wiki/The_Martyred_Pre...   
3  https://en.wikipedia.org/wiki/Terrible_Teddy,_...   
4  https://en.wikipedia.org/wiki/Jack_and_the_Bea...   

                                                Plot  \
0  A bartender is working at a saloon, serving dr...   
1  The moon, painted with a smiling face hangs ov...   
2  The film, just over a minute long, is composed...   
3  Lasting just 61 seconds and consisting of two ...   
4  The earliest known adaptation of the classic f...   

                                          clean_plot  
0  a bartender is working at a saloon, serving dr...  
1  the moon, painted with a smiling face hangs ov...  
2  the film, just over a minute long, is composed...  
3  lasting just 61 seconds and consisting of two ...  
4  the earliest known adaptation of the classic f...

In [69]:
df1['clean_plot'][0]

"a bartender is working at a saloon, serving drinks to customers. after he fills a stereotypically irish man's bucket with beer, carrie nation and her followers burst inside. they assault the irish man, pulling his hat over his eyes and then dumping the beer over his head. the group then begin wrecking the bar, smashing the fixtures, mirrors, and breaking the cash register. the bartender then sprays seltzer water in nation's face before a group of policemen appear and order everybody to leave."

In [70]:
df1["Director"] = df1["Director"].fillna("").str.strip()
df1["Cast"] = df1["Cast"].fillna("").str.strip()
df1["Genre"] = df1["Genre"].fillna("").str.strip()
    

In [71]:
df1.isna().sum()

Release Year        0
Title               0
Origin/Ethnicity    0
Director            0
Cast                0
Genre               0
Wiki Page           0
Plot                0
clean_plot          0
dtype: int64

In [72]:
def normalize_cast(cast):
    if not cast or cast.strip() == "":
        return []
    
    actors = cast.split(",")
    actors = [a.strip() for a in actors if a.strip()]
    return actors



In [73]:
def normalize_genre(genre):
    if not genre or genre.strip() == "":
        return []
    
    genres = genre.split(",")
    genres = [g.strip().title() for g in genres if g.strip()]
    return genres


In [74]:
df1["cast_list"] = df1["Cast"].apply(normalize_cast)
df1["genre_list"] = df1["Genre"].apply(normalize_genre)


In [76]:
df1.head(100)

Release Year                             Title Origin/Ethnicity  \
0           1901            Kansas Saloon Smashers         American   
1           1901     Love by the Light of the Moon         American   
2           1901           The Martyred Presidents         American   
3           1901  Terrible Teddy, the Grizzly King         American   
4           1902            Jack and the Beanstalk         American   
..           ...                               ...              ...   
95          1914                  The Property Man         American   
96          1914                        Recreation         American   
97          1914                Rose of the Rancho         American   
98          1914                      The Rounders         American   
99          1914                       Salomy Jane         American   

                              Director  \
0                                        
1                                        
2                                        
3                                        
4   George S. Fleming, Edwin S. Porter   
..                                 ...   
95                     Charlie Chaplin   
96                     Charles Chaplin   
97                    Cecil B. DeMille   
98                     Charlie Chaplin   
99                        William Nigh   

                                                 Cast    Genre  \
0                                                                
1                                                                
2                                                                
3                                                                
4                                                                
..                                                ...      ...   
95    Charlie Chaplin, Phyllis Allen, Alice Davenport   comedy   
96                                    Charles Chaplin            
97  Bessie Barriscale, Jane Darwell, Jeanie MacPhe...  western   
98     Charlie Chaplin, Fatty Arbuckle, Phyllis Allen   comedy   
99                    Beatriz Michelena, House Peters  western   

                                            Wiki Page  \
0   https://en.wikipedia.org/wiki/Kansas_Saloon_Sm...   
1   https://en.wikipedia.org/wiki/Love_by_the_Ligh...   
2   https://en.wikipedia.org/wiki/The_Martyred_Pre...   
3   https://en.wikipedia.org/wiki/Terrible_Teddy,_...   
4   https://en.wikipedia.org/wiki/Jack_and_the_Bea...   
..                                                ...   
95     https://en.wikipedia.org/wiki/The_Property_Man   
96    https://en.wikipedia.org/wiki/Recreation_(film)   
97   https://en.wikipedia.org/wiki/Rose_of_the_Rancho   
98  https://en.wikipedia.org/wiki/The_Rounders_(19...   
99  https://en.wikipedia.org/wiki/Salomy_Jane_(191...   

                                                 Plot  \
0   A bartender is working at a saloon, serving dr...   
1   The moon, painted with a smiling face hangs ov...   
2   The film, just over a minute long, is composed...   
3   Lasting just 61 seconds and consisting of two ...   
4   The earliest known adaptation of the classic f...   
..                                                ...   
95  Charlie is in charge of stage "props" and has ...   
96  Seated in a park, Charlie gives his expert att...   
97  Esra Kincaid (La Reno) takes land by force and...   
98  A drunk reveller (Chaplin) returns home to a s...   
99  Rough-and-tumble Gold Rush-era California: a w...   

                                           clean_plot  \
0   a bartender is working at a saloon, serving dr...   
1   the moon, painted with a smiling face hangs ov...   
2   the film, just over a minute long, is composed...   
3   lasting just 61 seconds and consisting of two ...   
4   the earliest known adaptation of the classic f...   
..                                                ...   
95  charlie is in charge of stage "props" and has ...   
96  seated in a park, charl

In [81]:
df1[["Title", "Cast", "cast_list"]].sample(5)
df1[["Title", "Genre", "genre_list"]].sample(5)


Title           Genre        genre_list
29101  Vipra Narayana Dubbed from Telugu[3]                                []
28365                    Ladies & Gentleman          comedy          [Comedy]
20306                          Silver Bears          comedy          [Comedy]
15182                  Daddy's Little Girls  drama, romance  [Drama, Romance]
2707                    When Tomorrow Comes           drama           [Drama]

In [85]:
(df1["cast_list"].apply(len) == 0).sum()
(df1["genre_list"].apply(len) == 0).sum()


6110

In [86]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34884 entries, 0 to 34883
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   Release Year      34884 non-null  int64 
 1   Title             34884 non-null  object
 2   Origin/Ethnicity  34884 non-null  object
 3   Director          34884 non-null  object
 4   Cast              34884 non-null  object
 5   Genre             34884 non-null  object
 6   Wiki Page         34884 non-null  object
 7   Plot              34884 non-null  object
 8   clean_plot        34884 non-null  object
 9   cast_list         34884 non-null  object
 10  genre_list        34884 non-null  object
dtypes: int64(1), object(10)
memory usage: 2.9+ MB
